### PR TITLE
[REVIEW] Port legacy hash partition kernel to new API

### PR DESCRIPTION
**Pull request by trevorsm7**
Closes https://github.com/rapidsai/cudf/issues/3566

Replaces https://github.com/rapidsai/cudf/pull/3636

### CODE REVIEW

**Comment by trevorsm7** [comment]
@nsakharnykh I ported the legacy hash partition kernel here so that we can apply the optimizations from https://github.com/rapidsai/cudf/pull/2623

**Comment by trevorsm7** [comment]
The LargeInput unit test that I wrote for my thrust-based implementation is failing for the legacy implementation. I observe the same result even with the legacy `gdf_hash_partition` API. This test calls hash partition twice with the same input and expects the result to be equal. The partition offsets are equal, so I think the values are still falling into the same partitions, but the ordering of the values within each partition may be non-deterministic.

If this is expected, I can rework the unit test to verify that values fall in the same partitions but not necessarily in the same order.

**Comment by trevorsm7** [comment]
The order of the rows within each partition is allowed to be non-deterministic as long as the rows end up in the same partition each time, so I updated the test accordingly.

File: cpp/src/hash/hashing.cu [hunk 1]
```diff
@@ -77,84 +78,276 @@ class bitwise_partitioner
...
+ * @param num_partitions THe number of partitions
+ * @param[out] block_partition_offsets Array that holds the offset of each partition for each thread block,
+ * i.e., { {block0 partition0 offset, block1 partition0 offset, ...}, 
+         {block0 partition1 offset, block1 partition1 offset, ...},
+         ...
+         {block0 partition(num_partitions-1) offset, block1 partition(num_partitions -1) offset, ...} }
+ */
+/* ----------------------------------------------------------------------------*/
+__global__ 
+void compute_row_output_locations(size_type * row_partition_numbers,
```
Comment (vuule) [inline cpp/src/hash/hashing.cu]: I suggest adding the restrict keyword to pointer kernel parameters:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#restrict
I've seen it improve performance of simple kernels by 5X (this one is far from simple, but it should still help).

File: cpp/src/hash/hashing.cu [hunk 2]
```diff
@@ -77,84 +78,276 @@ class bitwise_partitioner
...
+ * @param[out] global_partition_sizes The number of rows in each partition.
+ */
+/* ----------------------------------------------------------------------------*/
+template <class row_hasher_t, typename partitioner_type>
+__global__
+void compute_row_partition_numbers(row_hasher_t the_hasher,
+                                   const size_type num_rows,
+                                   const size_type num_partitions,
+                                   const partitioner_type the_partitioner,
+                                   size_type * row_partition_numbers,
```
Comment (vuule) [inline cpp/src/hash/hashing.cu]: could use restrict keyword here too

File: cpp/src/hash/hashing.cu [hunk 3]
```diff
@@ -77,84 +78,276 @@ class bitwise_partitioner
...
+      <<<grid_size, BLOCK_SIZE, num_partitions * sizeof(size_type), stream>>>
+          (row_output_locations, num_rows, num_partitions, scanned_block_partition_sizes);
+
+  auto scatter_map = column_view{data_type{INT32}, num_rows, row_output_locations};
+  auto output = experimental::detail::scatter(input, scatter_map, input, false, mr, stream);
+
+  return std::make_pair(std::move(output), std::move(partition_offsets));
 }
 
+struct nvtx_raii {
```
Comment (vuule) [inline cpp/src/hash/hashing.cu]: maybe add a comment to help understand what this is.

File: cpp/src/hash/hashing.cu (hunk 1 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Good idea, but it doesn't seem to make a difference here. Here are the benchmark results with and without adding `__restrict__`
Execution time for 32 cols per number of rows

Comment (vuule) [reply cpp/src/hash/hashing.cu]: :(
@trevorsm7 did you add the keyword to all pointer parameters?

Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: However, it didn't cause the unit tests to fail, so I suppose there's no harm in adding the keyword here anyway.

Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Yes, I added it to all there pointer parameters of `compute_row_output_locations` and of `compute_row_partition_numbers`

File: cpp/src/hash/hashing.cu (hunk 3 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Done

File: cpp/src/hash/hashing.cu (hunk 2 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Done

**Comment by revans2** [comment]
Is there anything else left to merge this in except for someone hitting the button?  Because it looks like it is ready to go and I would love to start using it.

**Pull request merged**